js_modules/dagster-ui/packages/ui-core/src/asset-data/AssetLiveDataThreadManager.tsx

```diff
@@ -0,0 +1,241 @@
+      (key) => {
+        delete this.lastFetchedOrRequested[key];
+      },
+    );
+  }
+
+  // Function used by threads.
+  public determineAssetsToFetch(keys: string[]) {
+    const assetsToFetch: AssetKeyInput[] = [];
+    const assetsWithoutData: AssetKeyInput[] = [];
+    while (keys.length && assetsWithoutData.length < BATCH_SIZE) {
+      const key = keys.shift()!;
+      const isRequested = !!this.lastFetchedOrRequested[key]?.requested;
+      if (isRequested) {
+        continue;
+      }
+      const lastFetchTime = this.lastFetchedOrRequested[key]?.fetched ?? null;
+      if (lastFetchTime !== null && Date.now() - lastFetchTime < this.pollRate) {
+        continue;
+      }
+      if (lastFetchTime && isDocumentVisible()) {
+        assetsToFetch.push({path: key.split('/')});
+      } else {
+        assetsWithoutData.push({path: key.split('/')});
+      }
+    }
+
+    // Prioritize fetching assets for which there is no data in the cache
+    return assetsWithoutData.concat(assetsToFetch).slice(0, BATCH_SIZE);
+  }
+
+  public areKeysRefreshing(assetKeys: AssetKeyInput[]) {
+    for (const key of assetKeys) {
+      const stringKey = tokenForAssetKey(key);
+      if (!this.lastFetchedOrRequested[stringKey]?.fetched) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private getAllObservedKeys() {
+    const threads = Object.values(this.threads);
+    return Array.from(new Set(threads.flatMap((thread) => thread.getObservedKeys())));
+  }
+
+  public getOldestDataTimestamp() {
+    const allAssetKeys = Object.keys(this.listeners).filter((key) => this.listeners[key]?.length);
+    let isRefreshing = allAssetKeys.length ? true : false;
+    let oldestDataTimestamp = Infinity;
+    for (const key of allAssetKeys) {
+      if (this.lastFetchedOrRequested[key]?.fetched) {
+        isRefreshing = false;
+      }
+      oldestDataTimestamp = Math.min(
+        oldestDataTimestamp,
+        this.lastFetchedOrRequested[key]?.fetched ?? Infinity,
+      );
+    }
+    return {
+      isRefreshing,
+      oldestDataTimestamp: oldestDataTimestamp === Infinity ? 0 : oldestDataTimestamp,
+    };
+  }
+
+  public _updateCache(data: Record<string, LiveDataForNode>) {
+    Object.assign(this.cache, data);
+  }
+
+  public onDocumentVisiblityChange(isDocumentVisible: boolean) {
+    if (isDocumentVisible) {
+      if (this.isPaused) {
+        this.unpause();
+      }
+    } else if (!this.isPaused) {
+      this.pause();
+    }
+  }
+
+  private pause() {
+    this.isPaused = true;
+    Object.values(this.threads).forEach((thread) => {
+      thread.stopFetchLoop();
+    });
+  }
+
+  private unpause() {
+    this.isPaused = false;
+    Object.values(this.threads).forEach((thread) => {
+      thread.startFetchLoop();
+    });
+  }
+
+  public getCacheEntry(key: AssetKeyInput) {
+    return this.cache[tokenForAssetKey(key)];
+  }
+
+  public _markAssetsRequested(assetKeys: AssetKeyInput[]) {
```

Comment:
Ah man `public _` is a bummer, I wish TS had `fileprivate` but it looks like this is really the best option 😅

Reply:
I could have used protected as "filePrivate" but really the reason it's public with an underscore is because it's used by the thread's, I suppose I could inject these functions into the threads instead while keep this method private but something about that I didn't like though maybe it's the right thing to do to keep public interface tidy